### PR TITLE
add a broadcaster for range sensor

### PR DESCRIPTION
**Pull request by flochre**
Hello!

I have been working on a broadcaster for Range sensors.
Here is a Pull Request to integrate it!

In the file ros2_controllers/range_sensor_broadcaster/include/range_sensor_broadcaster/range_sensor_broadcaster.hpp I have not put my name since I basically just change IMU with Range.. Is this ok or should I put my name on it?

I have prepared the following file ros2_control/controller_interface/include/semantic_components/range_sensor.hpp
on my fork https://github.com/flochre/ros2_control/blob/feature/add-range-sensor-broadcast/controller_interface/include/semantic_components/range_sensor.hpp - Should I already make a pull request by ros2_control?

Thanks in advance,

Cheers!

Contributions via pull requests are much appreciated. Before sending us a pull request, please ensure that:

1. Limited scope. Your PR should do one thing or one set of things. Avoid adding “random fixes” to PRs. Put those on separate PRs.
2. Give your PR a descriptive title. Add a short summary, if required.
3. Make sure the pipeline is green.
4. Don’t be afraid to request reviews from maintainers.
5. New code = new tests. If you are adding new functionality, always make sure to add some tests exercising the code and serving as live documentation of your original intention.

To send us a pull request, please:

- [x] Fork the repository.
- [x] Modify the source; please focus on the specific change you are contributing. If you also reformat all the code, it will be hard for us to focus on your change.
- [x] Ensure local tests pass. (`colcon test` and `pre-commit run` (requires you to install pre-commit by `pip3 install pre-commit`)
- [x] Commit to your fork using clear commit messages.
- [x] Send a pull request, answering any default questions in the pull request interface.
- [ ] Pay attention to any automated CI failures reported in the pull request, and stay involved in the conversation.

### CODE REVIEW

**Comment by bmagyar** [comment]
Please check the results from the CI. The format stage is failing, you need to run `pre-commit run --all` to let it fix things for you. Also check the other stages. The code currently doesn't compile and you also have a lot of conversion warnings (purple text)

File: range_sensor_broadcaster/package.xml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>range_sensor_broadcaster</name>
+  <version>0.0.1</version>
```
[the remark this answers is not in the document]
Comment (flochre) [reply range_sensor_broadcaster/package.xml]: Done

**Comment by flochre** [comment]
The compilation is also not working because of the requirement to ros2_control
`repositories:
    ros2_control:
        type: git
        url: https://github.com/flochre/ros2_control.git
        version: release/rolling-ruediger2-v1.0
`

**Comment by flochre** [comment]
About the conversion warning, I don't get them for the range_sensor_broadcaster
Those I have are coming from other packages.. Should I still correct them?

**Comment by bmagyar** [comment]
Well let's start at the compilation first.
Could you please open a PR against `ros2_control` then? We can discuss here all we want if we are missing the other side ;)

**Comment by flochre** [comment]
I think everything should be set now ;)

**Comment by flochre** [comment]
Nice, compiling and testing worked!
Should I update the branch or you do it? Because if I do it the git actions will need to run again?
What do you prefere?
Thanks!

File: range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  init_broadcaster("test_range_sensor_broadcaster");
+
+  range_broadcaster_->on_configure(rclcpp_lifecycle::State());
+  range_broadcaster_->on_activate(rclcpp_lifecycle::State());
+
+  sensor_msgs::msg::Range range_msg;
+  subscribe_and_get_message(range_msg);
+
+  EXPECT_EQ(range_msg.header.frame_id, frame_id_);
+  EXPECT_PRED_FORMAT2(::testing::FloatLE, range_msg.range, sensor_range_);
```
Comment (bmagyar) [inline range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp:172]:
```suggestion
  EXPECT_THAT(range_msg.range, ::testing::Le(sensor_range_));
```
is this accurate to what you meant here?

Comment (flochre) [reply range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp:172]: Ok, yes I guess you are right this is the same.
In the mean time I read the [matchers.md](https://github.com/google/googletest/blob/main/docs/reference/matchers.md)

I what I really want to test is the following :

`EXPECT_THAT(range_msg.range, ::testing::FloatEq(sensor_range_));`

So I will change all the occurrences and commit again + rebase

Comment (flochre) [reply range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp]: Do you agree with the changes?

File: range_sensor_broadcaster/package.xml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>range_sensor_broadcaster</name>
+  <version>3.12.0</version>
```
Comment (bmagyar) [inline range_sensor_broadcaster/package.xml:5]:
```suggestion
  <version>3.14.0</version>
```

File: range_sensor_broadcaster/package.xml [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>range_sensor_broadcaster</name>
+  <version>3.14.0</version>
+  <description>Controller to publish readings of Range sensors.</description>
+  <maintainer email="flopubs@gmail.com">flochre</maintainer>
```
Comment (bmagyar) [inline range_sensor_broadcaster/package.xml:7]:
```suggestion
  <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
  <author email="flopubs@gmail.com">flochre</author>
```

Perhaps you want to use your full name here @flochre ?

File: range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  range_broadcaster_ = std::make_unique<range_sensor_broadcaster::RangeSensorBroadcaster>();
+}
+
+void RangeSensorBroadcasterTest::TearDown() { range_broadcaster_.reset(nullptr); }
+
+controller_interface::return_type RangeSensorBroadcasterTest::init_broadcaster(
+  std::string broadcaster_name)
+{
+  controller_interface::return_type result = controller_interface::return_type::ERROR;
+  // result = range_broadcaster_->init("test_range_sensor_broadcaster");
```
Comment (bmagyar) [inline range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp:37]: Do we need this line?

File: range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+      break;
+    }
+  }
+  ASSERT_GE(max_sub_check_loop_count, 0) << "Test was unable to publish a message through "
+                                            "controller/broadcaster update loop";
+
+  // take message from subscription
+  rclcpp::MessageInfo msg_info;
+  subscription->take(range_msg, msg_info);
+  // ASSERT_TRUE(subscription->take(range_msg, msg_info));
```
Comment (bmagyar) [inline range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp:91]: Same here

**Review by bmagyar — CHANGES_REQUESTED** [review]
I didn't notice the load controller test here. Would you mind adding that too? Just copy it from any controller

File: range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp (hunk 5 above)
Comment (flochre) [reply range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp]: I deleted this one

File: range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp (hunk 6 above)
Comment (flochre) [reply range_sensor_broadcaster/test/test_range_sensor_broadcaster.cpp]: I took this one back

**Comment by flochre** [comment]
Load test as been added and all test are passing in my workspace

**Comment by flochre** [comment]
Is there anything more I should do?

File: range_sensor_broadcaster/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    ${CMAKE_CURRENT_SOURCE_DIR}/test/range_sensor_broadcaster_params.yaml)
+  target_include_directories(test_range_sensor_broadcaster PRIVATE include)
+  target_link_libraries(test_range_sensor_broadcaster
+    range_sensor_broadcaster
+  )
+  ament_target_dependencies(test_range_sensor_broadcaster
+    hardware_interface
+  )
+
+
```
Comment (bmagyar) [inline range_sensor_broadcaster/CMakeLists.txt:86]:
```suggestion
```

File: range_sensor_broadcaster/package.xml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  <depend>hardware_interface</depend>
+  <depend>pluginlib</depend>
+  <depend>rclcpp</depend>
+  <depend>rclcpp_lifecycle</depend>
+  <depend>realtime_tools</depend>
+  <depend>sensor_msgs</depend>
+
+  <test_depend>ament_cmake_gmock</test_depend>
+  <test_depend>ament_lint_auto</test_depend>
+  <test_depend>ament_lint_common</test_depend>
```
Comment (bmagyar) [inline range_sensor_broadcaster/package.xml:25]:
```suggestion
```

**Review by bmagyar — APPROVED** [review]
Thank you for sticking through the review, will wait for the CI to turn green and merge it! 
I'll also backport this PR to humble

**Comment by christophfroehlich** [comment]
And I already got a notification of a warning in the docs ;)

**Pull request merged**